### PR TITLE
Fix dataloader hang with large sampler

### LINKED ISSUE #48666: [dataloader] hang on python exit when has iter ref and sampler yields large indices
## 🐛 Bug

Using the DataLoader causes the main process to hang on exit waiting to join the indices queue putting thread when the sampler is large (e.g., when `batch_size` is large) and a reference to the iterator is kept around (e.g., when `persistent_workers=True`). For instance, see the following simple script:

```py
import torch

class Dataset(torch.utils.data.Dataset):
    def __len__(self):
        return 10000000

    def __getitem__(self, any):
        return torch.empty(0)

if __name__ == '__main__':
    dl = torch.utils.data.DataLoader(
        Dataset(),
        batch_size=40960,
        num_workers=1)

    it = iter(dl)

    for i, x in enumerate(it):
        print(x.shape)
        raise RuntimeError()
```

Running the above in the below environment (pretty standard) will hang at exiting. Some probing suggests that this is due to the main process try to join the `_index_queues`. And, **due to** the iterator being around, the ordering of python atexit calls decides to call the `multiprocessing.Queue` clean-up (which joins the queue putting thread) before cleaning up `_MultiprocessingDataLoaderIter.__del__` (which would clean-up nicely). 

Finally, since the above uses a sampler with *large* indices sent to workers at each time (constructed by the dataloader using a `RandomSampler` wrapped in a `BatchSampler`), the putting thread is hanging waiting for more room in the pipe. The stderr will hang at 
```py
torch.Size([40960, 0])
Traceback (most recent call last):
  File "XXXX/test.py", line 34, in <module>
    raise RuntimeError()
RuntimeError
```

NB: the worker yields empty tensors, and thus its putter thread shouldn't be the reason of hanging.

To confirm that this also happens with a simpler sampler, the following script with disabled auto-batching hangs similarly:
```py
import torch


class BigSampler(object):
    def __iter__(self):
        for idx in range(20):
            yield [idx for idx in range(40960)]


class Dataset(torch.utils.data.Dataset):

    def __init__(self):
        self.first_time = True

    def __len__(self):
        return 10000000

    def __getitem__(self, indices):
        # NB: this is called *once* per dataloader fetch since we disabled auto-batching!
        if not self.first_time:
            # simulate some work so that the index_queue is likely non-empty 
            # (and thus the main process's queue putting is waiting). 
            import time
            time.sleep(0.1)
        self.first_time = False
        return torch.empty(0)


if __name__ == '__main__':
    dl = torch.utils.data.DataLoader(
        Dataset(),
        batch_size=None,
        sampler=BigSampler(),
        num_workers=1)

    it = iter(dl)

    for i, x in enumerate(it):
        print(x.shape)
        raise RuntimeError()
```

## Environment
```
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 16.04.7 LTS (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.14.0

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration:
GPU 0: TITAN V
GPU 1: TITAN V
GPU 2: TITAN V
GPU 3: TITAN V

Nvidia driver version: 455.23.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip] numpy==1.17.2
[pip] torch==1.7.0
[pip] torchaudio==0.7.0a0+ac17b64
[pip] torchfile==0.1.0
[pip] torchvision==0.8.1
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.1.243             h6bb024c_0
[conda] mkl                       2019.4                      243
[conda] mkl-include               2019.4                      243
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.2.0            py37h23d657b_0
[conda] mkl_random                1.1.0            py37hd6b4f25_0
[conda] numpy                     1.16.0                   pypi_0    pypi
[conda] numpy-base                1.17.2           py37hde5b4d6_0
[conda] pytorch                   1.7.0           py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torch                     1.3.0                    pypi_0    pypi
[conda] torchaudio                0.7.0                      py37    pytorch
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchvision               0.6.0a0+35d732a          pypi_0    pypi
```

cc @SsnL @VitalyFedyunin @ejguan

### PR DESCRIPTION
**Pull request by ssnl**
Fixes https://github.com/pytorch/pytorch/issues/48666

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -867,7 +896,9 @@ def __init__(self, loader):
         for i in range(self._num_workers):
             # No certainty which module multiprocessing_context is
             index_queue = multiprocessing_context.Queue()  # type: ignore
-            # index_queue.cancel_join_thread()
+            # Need to `cancel_join_thread` here!
+            # See sections (2) and (3b) above.
+            index_queue.cancel_join_thread()
```
Comment (SsnL) [inline torch/utils/data/dataloader.py:901]: this is the only relevant code change. the rest is just doc and test

**Review by VitalyFedyunin — APPROVED** [review]
> The code looks good, but how slow is the test?
Nevermind, it is joined right away.

**Comment by VitalyFedyunin** [comment]
> The code looks good, but how slow is the test?

Ah, nevermind, we are joining right away.

**Comment by SsnL** [comment]
@VitalyFedyunin 

It runs in 3s on my linux box
```
> python ../pytorch/test/run_test.py -i test_dataloader -v  -- Test
der.test_large_sampler_indices
Fail to import hypothesis in common_utils, tests are not derandomized
Test executor: [XXXXXXXXXXXXXXXXXXXX]
Selected tests: test_dataloader
Running test_dataloader ... [2020-12-01 17:37:46.778711]
Fail to import hypothesis in common_utils, tests are not derandomized
test_large_sampler_indices (__main__.TestDataLoader) ... Fail to import hypothesis in common_utils, tests are not derandomized
Fail to import hypothesis in common_utils, tests are not derandomized
ok

----------------------------------------------------------------------
Ran 1 test in 3.528s

OK
```

**Pull request closed without merging**